### PR TITLE
Feature/#10 related posts block

### LINKED ISSUE #10: Related Posts Block
Duplicate our current Related Post Block functionality as a Gutenblock.

http://gutenberg.wdslab.com/acf-blocks-page/

**Requirements**

- [x] Background Image, Video, and Color
- [x] Title
- [ ] Manual Post Selector (up to 3)
- [x] Font Color
- [x] Custom CSS Class

**Screenshot**

screenshot

### PR DESCRIPTION
**Pull request by jomurgel**
Closes #10

### DESCRIPTION ###
Adds related posts selection block which has the following features.
- Paginated post retrieval
- Integrated search component (added to global components)
- Selection and removal of wanted posts.
- Off-focus render of post grid (3-column grid)
- Background color/video/image options.
- Text options
- Advanced options (CSS animation + custom class)
- Loading animation when posts are loading (added to global components).
- Added accessibility considerations

Phase 2 (roadmap below) should address some styling updates once #40 has been merged into master.

### SCREENSHOTS ###
kapture 2018-04-17 at 9 05 56

kapture 2018-04-17 at 9 23 15

screenshot 2018-04-17 09 06 11

### OTHER ###
- [x] Is this issue accessible? (Section 508/WCAG 2.0AA)
- [x] Does this issue pass linting? (ESLint)

### STEPS TO VERIFY ###
Add block (see screen cap above)

### DOCUMENTATION ###
https://github.com/WebDevStudios/wds-gutenberg/wiki/Global-Components

https://github.com/WebDevStudios/wds-gutenberg/wiki/Blocks

I may need to make updates here, I'm not totally sure what information we want to include here.

### ROADMAP ###
- [ ] Support for Custom Post Types + Taxonomies.
- [ ] Support max post count + metabox notice if trying to add more.
- [ ] Add meta box notice if post trying to save 0 selected posts.
- [ ] Post selection limits.
- [ ] Update styling considerations once #40 is merged into master.

### CODE REVIEW

**Comment by coreymcollins** [comment]
I tried breaking this a bit by saving no posts, adding multiple blocks to the same page, messing around with block titles, etc etc. -- so far, I haven't been able to!

I noticed a couple of markup issues that I think we could probably resolve as part of this PR, those being that the image and post title are not linked to the single post.  Our current ACF blocks link both of those items.

The functionality of the block itself is working for me, but I am seeing an interesting situation when you update a post after it has been added to a block.

For example -- I created these two instances of Related Posts with the same two posts.  I saved my page, and then I went back and edited "This is a post" by changing the post title to "This is a post!" and setting a feature image.  I then edited the first instance of Related Posts on my page my adding and removing the post and saved my page.  The output is now this:


So, the data saved in the Related Posts block is a picture of the posts at that exact moment in time.  It isn't enough to just re-save the page where the Related Posts block, because the saved post in the component also reflects the old data:


I think we may need to just store the IDs of the saved posts and then either adjust the output here to grab our post content by ID rather than storing the values associated with the post.  Maybe this is a use-case for a PHP render so we can grab everything with core WP functions, rather than something like running a WP-API query for each individual post ID?  I'm not sure what the performance comparison would be between those two options.

**Comment by coreymcollins** [comment]
Couple other edge-case notes.  If I create a block with just a block title and no selected posts, the markup for the posts container and UL still display:

**Comment by jomurgel** [comment]
EOD Update here.

I've converted this block to use the php render callback function. Since our attributes object had our posts array as stringified JSON, that was easy to output to php.

Ended up being able to completely remove the post objects from the attribute and create an array of IDs for use within the render callback function.

Inside the component, I fed the "selected posts" on `componentDidMount` in order to rehydrate the block upona reloading.

This update has caused an errro which I still need to look into.

> Warning: Cannot update during an existing state transition (such as within `render` or another component's constructor). Render methods should be a pure function of props and state; constructor side-effects are an anti-pattern, but can be moved to `componentWillMount`.

Which is being caused by the onChange handler of the selected posts output component.

``` javascript
<Output
	activeClass={ this.state }
	title="Selected Posts"
	className="related-right-column"
	key="related-right-column"
	posts={ this.state.selectedPosts }
	handleEvent={ this.handleEvent }
	onChange={ this.props.setAttributes( {
		selectedPostsJSON: JSON.stringify( this.state.selectedPosts.map( post => post.id ) ),
		selectedPosts: this.state.selectedPosts,
	} ) }
/>
```

So now the posts in the block and on the front end will update on refresh if a post or post content has changed.

May also need to address a few style updates to the front end once all the code is error-free.

**Comment by jomurgel** [comment]
Actually just made an update to the onChange handler in the Output component noted above.

This
``` javascript
onChange={ this.props.setAttributes( {
	selectedPostsJSON: JSON.stringify( this.state.selectedPosts.map( post => post.id ) ),
	selectedPosts: this.state.selectedPosts,
} ) }
```

Becomes this
``` javascript
onChange={ this.props.setAttributes.bind( {
	selectedPostsJSON: JSON.stringify( this.state.selectedPosts.map( post => post.id ) ),
	selectedPosts: this.state.selectedPosts,
} ) }
```

By using `bind()`, the function won’t be called during the state transition anymore, it will complete the process and leave the function to be called later.

Reference: http://4dev.tech/2016/03/reactjs-error-cannot-update-during-an-existing-state-transition-such-as-within-render-render-methods-should-be-a-pure-function-of-props-and-state/

**Comment by jomurgel** [comment]
So the update above had no positive effect, but a solution was found with Eric's help.

## Changes
- Refactors function to utilize state 100% and set state on click event.
- Removes `onChange` handler from Selected `Output` component (behavior handled inside the `handleEvent` function.
- Calls `setAttributes` inside the `handleEvent` function.
- Reworks the `post-render.js` `withAPIData` function to utilize state — needs to grab ids now.
- Fixed the output of images in the post renderer.
- Fixes scrolling in the post container ul if focus away — adds `componentDidUpdate` and `componentWillUnmount` to prevent errors and retain scrolling functionality

## Description
These updates just cleans things up and relies on state instead of partial state partial attributes. If a post is updated the front end will reflect the change in title/image/excerpt and the post titles/content will update in the block on page refresh.

screenshot 2018-04-20 09 16 11

screenshot 2018-04-20 10 25 09

screenshot 2018-04-20 10 25 17

**Comment by coreymcollins** [comment]
Testing this out this morning under three conditions:
- Block with title and posts selected
- Block with title and no posts selected
- Block with no title and posts selected

Notes below...

Despite selecting the posts in a specific order, they are being displayed in the order they are found in the results (left column) rather than the added (right column):



----------

If I don't select any posts, I still get posts displaying:



This one may be an edge case because, why would someone add the block without selecting posts?  But, this was working almost as expected before the latest updates although it was putting out an empty div.

Is this done on purpose?  I tested publishing a new post and after doing so and refreshing the page on the frontend, the new post displays:


----------

I did test editing a post and can confirm that its changes now reflect on the frontend when refreshing the page, so thank you for taking care of that!  This also takes care of updating the title/image/excerpt in the dashboard as well, so double thanks!

File: src/blocks/related-posts/render.php [hunk 1]
```diff
@@ -1,43 +1,127 @@
...
+					<li <?php post_class( 'column' ); ?> tabindex="0">
+						<?php if ( has_post_thumbnail() ) : ?>
+							<?php echo wp_get_attachment_image( $post_thumb_id, 'medium_large' ); ?>
+						<?php endif; ?>
+						<h3 class="h1">
+							<a href="<?php the_permalink(); ?>">
+								<?php the_title(); ?>
+							</a>
+						</h3>
+						<div class="">
```
Comment (coreymcollins) [inline src/blocks/related-posts/render.php]: Can we add a class here or remove `class=""` if we're not going to use a class at all?

**Comment by coreymcollins** [comment]
@gregrickaby -- going to pass this over to you for a final once-over before merging in.  Thanks!  🎉

**Pull request merged**